File: PatchInstaller/Module1.vb
Imports System.Runtime.CompilerServices
Imports Patch_Management
Imports Patch_Management.WindowsPatchManagement

Module Module1

    Sub Main(ByVal args() As String)
        Try
            Dim DriverInstall As Boolean = False
            Dim RebootApproved As Boolean = False
            Dim SoftwareUpdates As Boolean = False

            If args.Count = 0 Then
                Help()
                Exit Sub
            End If

            For Each arg In args
                If arg = "?" Or arg.ToLower = "help" Then
                    Help()
                    Exit Sub
                ElseIf arg = "-drivers" Then
                    DriverInstall = True
                ElseIf arg = "-reboot" Then
                    RebootApproved = True
                ElseIf arg = "-updates" Then
                    SoftwareUpdates = True
                ElseIf arg.Equals("history") Then
                    Console.WriteLine("Install Date,HResult Code,HResult Description,Revision,Category,Title,Description")
                    For Each WUpdate In GetUpdateHistory()
                        Console.WriteLine(WUpdate.InstallDate & "," & WUpdate.InstallResult.ToString("X") & "," & WindowsPatchManagement.GetHCODEDescription(WUpdate.InstallResult) & "," & WUpdate.Revision & "," & WUpdate.Category & "," & WUpdate.Title & "," & WUpdate.Description)
                    Next
                    Exit Sub
                Else
                    Console.WriteLine("Invalid Argument: " & arg)
                    Exit Sub
                End If
            Next

            Dim rebootrequired As Boolean = InstallUpdates(DriverInstall, SoftwareUpdates)
            If rebootrequired Then
                Console.WriteLine("Reboot required for updates.")
                If RebootApproved Then
                    'Reboot system.
                    Dim rebproc As New Process
                    rebproc.StartInfo.FileName = "C:\Windows\System32\shutdown.exe"
                    rebproc.StartInfo.Arguments = "/r /t 1"
                    rebproc.Start()
                End If
            End If

        Catch ex As Exception
            Console.WriteLine("Error installing updates: " & ex.ToString)
            Environment.ExitCode = 1
        End Try

        'Potential speed improvement using offline database of updates:  wsusscn2.cab
        'https://stackoverflow.com/questions/27337433/slow-wua-windows-update-api

        'https://support.microsoft.com/en-au/topic/a-new-version-of-the-windows-update-offline-scan-file-wsusscn2-cab-is-available-for-advanced-users-fe433f4d-44f4-28e3-88c5-5b22329c0a08


    End Sub

    Sub Help()
        Console.WriteLine("Patch Installer - Mitchell Hayden")
        Console.WriteLine("Installs windows updates.")
        Console.WriteLine("Arguments available:")
        Console.WriteLine("-updates     Include Windows Software Updates.")
        Console.WriteLine("-drivers     Include Hardware driver updates.")
        Console.WriteLine("-reboot      Reboot if required when updates completed.")
        Console.WriteLine("history      Show Windows update history for device.")
    End Sub

End Module

File: PatchInstaller/Module1.cs
﻿using Patch_Management;
using System;
using System.Diagnostics;
using System.Linq;
using static Patch_Management.WindowsPatchManagement;

namespace PatchInstaller
{

    static class Module1
    {

        public static void Main(string[] args)
        {

            try
            {
                if (args.Count() == 0)
                {
                    Help();
                    return;
                }

                Console.WriteLine("Patch Management - Mitchell Hayden");

                bool DriverInstall = false;
                bool RebootApproved = false;
                bool SoftwareUpdates = false;
                bool PreviewUpdates = false;
                bool scriptMode = false;

                //Arguments that can be combined.
                foreach (string arg in args)
                {
                    //single commands run and execute.

                    if (arg.Equals("?") | arg.ToLower().Equals("help"))
                    {
                        Help();
                        return;
                    }
                    else if (arg.Equals("history"))
                    {
                        WUpdateHistory.DisplayHistory();
                        return;
                    }
                    else if (arg.Equals("check"))
                    {
                        WUpdate.DisplayPendingUpdates();
                        return;
                    }
                    else if (arg.Equals("health"))
                    {
                        HealthCheck.GetWindowsUpdateLog(@"C:\Windows\Temp\PatchManagement\WindowsUpdate.log");
                        return;
                    }

                    //Combine commands.

                    else if (arg.Equals("-update"))
                    {
                        SoftwareUpdates = true;
                    }
                    else if (arg.Equals("-drivers"))
                    {
                        DriverInstall = true;
                    }
                    else if (arg.Equals("-reboot"))
                    {
                        RebootApproved = true;
                    }
                    else if (arg.Equals("-select"))
                    {
                        int index = int.Parse(args[1]);
                        InstallUpdate(index);
                        return;
                    }
                    else if(arg.Equals("-preview"))
                    {
                        PreviewUpdates = true;
                    }
                    else if(arg.Equals("-script"))
                    {
                        scriptMode = true;
                    }

                    //Invalid arguments.
                    else
                    {
                        Console.WriteLine("Invalid Argument: " + arg);
                        return;
                    }
                }

                if (RebootManager.CheckPendingRestart())
                {
                    Console.WriteLine("A pending restart has been detected.");
                }
                else
                {
                    Console.WriteLine("No pending restarts have been detected.");
                }
                Console.WriteLine();


                if(SoftwareUpdates || DriverInstall)
                {
                    if (InstallUpdates(DriverInstall, SoftwareUpdates, PreviewUpdates, scriptMode))
                    {
                        Console.WriteLine("Reboot required for updates.");
                        if (RebootApproved)
                        {
                            Console.WriteLine("Restarting System");
                            RebootManager.Restart();
                            return;
                        }
                    }
                    else
                    {
                        Console.WriteLine("Install complete.");
                    }
                }

                //Check for PendingReboot Flags.
                if (RebootApproved && RebootManager.CheckPendingRestart())
                {
                    Console.WriteLine("System is pending reboot - restarting");
                    RebootManager.Restart();
                    return;
                }


            }
            catch (Exception ex)
            {
                Console.WriteLine("Error installing updates: " + ex.ToString());
                Environment.ExitCode = 1;
            }

            // Potential speed improvement using offline database of updates:  wsusscn2.cab
            // https://stackoverflow.com/questions/27337433/slow-wua-windows-update-api

            // https://support.microsoft.com/en-au/topic/a-new-version-of-the-windows-update-offline-scan-file-wsusscn2-cab-is-available-for-advanced-users-fe433f4d-44f4-28e3-88c5-5b22329c0a08


        }



        static void Help()
        {
            Console.WriteLine("Patch Installer Help");
            Console.WriteLine("Install parameters available:");
            Console.WriteLine("-update      Include Windows Software Updates.");
            Console.WriteLine("-drivers     Include Hardware driver updates.");
            Console.WriteLine("-preview     Include Preview Updates.");
            Console.WriteLine("-reboot      Reboot if required when updates completed.");
            Console.WriteLine("-select x    Install single update where X is the index number available in the 'check' list.");
            Console.WriteLine("-script      Reduced status info in console output for remote script execution.");

            Console.WriteLine();
            Console.WriteLine("Reporting commands available:");
            Console.WriteLine("history      Show Windows update history for device.");
            Console.WriteLine("check        Check for available updates and show list.");
            Console.WriteLine("health       Writes windows update logs to disk and checks for errors.");
        }

    }
}